### PR TITLE
make relay starting stricter

### PR DESCRIPTION
**Pull request by evanmcc**
see the comment in the code.  pretty sure this busts the tests, I'll work on fixing up the mecking tomorrow.

### CODE REVIEW

File: src/relay/libp2p_relay_server.erl [hunk 1]
```diff
@@ -225,7 +225,17 @@ init_relay(#state{tid = TID, banlist = Banlist}) ->
...
-                                        libp2p_peer:has_public_ip(P) andalso
+                                        %% this is complex and restrictive, but there's a hard to
+                                        %% track down bug in the relay logic that sometimes allows
+                                        %% ipv4 addresses to be listed within the listen addrs.
+                                        %% this can cause relay establishment problems because the
+                                        %% relays only work if they're in the correct order, which
+                                        %% is impossible to guarantee.  so we work around by
+                                        %% narrowing the relay pool to stuff that we know is likely
+                                        %% to work, or to at least fail immediately.
+                                        ListenAddrs = libp2p_peer:listen_addrs(P),
```
Comment (vihu) [inline src/relay/libp2p_relay_server.erl:236]: Do we no longer need to check whether the peer `has_public_ip`?

File: src/relay/libp2p_relay_server.erl [hunk 2]
```diff
@@ -225,7 +225,17 @@ init_relay(#state{tid = TID, banlist = Banlist}) ->
...
+                                        length(ListenAddrs) == 1 andalso
+                                            libp2p_transport_tcp:is_public(hd(ListenAddrs)) andalso
                                             not libp2p_peer:is_stale(P, timer:minutes(45))
```
Comment (vihu) [inline src/relay/libp2p_relay_server.erl:239]: Probably better to hoist the timer value somewhere? App config maybe?

File: src/relay/libp2p_relay_server.erl (hunk 1 above)
Comment (evanmcc) [reply src/relay/libp2p_relay_server.erl:236]: we still check it

File: src/relay/libp2p_relay_server.erl (hunk 2 above)
Comment (evanmcc) [reply src/relay/libp2p_relay_server.erl:239]: yeah

**Review by macpie — APPROVED** [review]

**Comment by evanmcc** [comment]
old code:
```
(miner@127.0.0.1)35> maps:fold(fun(_K, V, Acc) -> maps:update_with(V, fun(X) -> X + 1 end, 1, Acc) end, #{}, maps:from_list([ {element(2, erl_angry_purple_tiger:animal_name( libp2p_crypto:bin_to_b58( X))),  Y} || {X, Y} <- [ {Address, libp2p_peer:has_public_ip(Peer)} || {Address, {ok, Peer}} <- [{Address, libp2p_peerbook:get(libp2p_swarm:peerbook(blockchain_swarm:swarm()), Address)} || Address <- libp2p_peerbook:keys(libp2p_swarm:peerbook(blockchain_swarm:swarm()))]], Y /= <<>>  ])). 
#{false => 8595,true => 5452}
```

new code:
```
maps:fold(fun(_K, V, Acc) -> maps:update_with(V, fun(X) -> X + 1 end, 1, Acc) end, #{}, maps:from_list([ {element(2, erl_angry_purple_tiger:animal_name( libp2p_crypto:bin_to_b58( X))),  Y} || {X, Y} <- [ {Address, begin LA = libp2p_peer:listen_addrs(Peer), length(LA) == 1 andalso libp2p_transport_tcp:is_public(hd(LA)) end} || {Address, {ok, Peer}} <- [{Address, libp2p_peerbook:get(libp2p_swarm:peerbook(blockchain_swarm:swarm()), Address)} || Address <- libp2p_peerbook:keys(libp2p_swarm:peerbook(blockchain_swarm:swarm()))]], Y /= <<>>  ])). 
#{false => 11737,true => 2310}
```

**Comment by evanmcc** [comment]
so this is not without cost, there's some concerns that there won't be enough relays if we enable this without fixing the underlying issue.

**Comment by evanmcc** [comment]
closing in favor of #309

**Pull request closed without merging**